src/main/java/org/jboss/remoting3/EndpointImpl.java

```diff
@@ -414,11 +422,17 @@ public void handleFailed(final IOException exception, final Void attachment) {
                     public void handleDone(final ConnectionHandlerFactory connHandlerFactory, final Void attachment) {
                         log.logf(getClass().getName(), Logger.Level.TRACE, null, "Registered successful result %s", connHandlerFactory);
                         final ConnectionImpl connection = new ConnectionImpl(EndpointImpl.this, connHandlerFactory, connectionProviderContext, destination);
-                        connections.add(connection);
                         connection.getConnectionHandler().addCloseHandler(SpiUtils.asyncClosingCloseHandler(connection));
                         connection.addCloseHandler(resourceCloseHandler);
                         connection.addCloseHandler(connectionCloseHandler);
-                        futureResult.setResult(connection);
+                        // see if we were closed in the meantime
+                        if (EndpointImpl.this.isCloseFlagSet()) {
```

Comment:
It is possible for this flag to change to "true" after this "if" statement but before the subsequent "setResult".